core/context.go

```diff
@@ -0,0 +1,222 @@
+// Copyright 2021 Solus Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+// #include <pwd.h>
+// #include <grp.h>
+// #include <unistd.h>
+import "C"
+
+import (
+	"fmt"
+	log "github.com/DataDrake/waterlog"
+	"os/exec"
+	gouser "os/user"
+	"strconv"
+)
+
+const minimumUID = 1000
+const wheelGroup = "sudo"
+
+// Context contains contextual system data, such as groups, users, and active shells, along with paths to certain binaries
+type Context struct {
+	users  []User
+	groups []gouser.Group
+	shells []string
+}
+
+// User is a convenience struct with information on a user entry in /etc/passwd
+type User struct {
+	Name     string
+	Groups   []string
+	IsActive bool
+	IsRoot   bool
+	IsAdmin  bool
+}
+
+// NewContext creates an initialized instance of a Context object
+func NewContext() (ctx *Context, err error) {
+	ctx = &Context{}
+	log.Debugln("Gathering system info...")
+
+	ctx.shells = activeShells()
+	log.Debugln("\tGathered active shells from /etc/shells")
+
+	if err = ctx.populateGroups(); err != nil {
+		return ctx, fmt.Errorf("failed to obtain groups from /etc/groups: %s", err)
+	}
+	log.Debugln("\tGathered groups from /etc/groups")
+
+	if err = ctx.populateUsers(); err != nil {
+		return ctx, fmt.Errorf("failed to obtain users from /etc/passwd: %s", err)
+	}
+	log.Debugln("\tGathered users from /etc/passwd")
+
+	return ctx, err
+}
+
+// FilterUsers returns users that fit into one or more of the given filters
+func (c *Context) FilterUsers(filters ...string) (filtered []User) {
+	for _, it := range c.users {
+		switch {
+		case contains(filters, "all"):
+			fallthrough
+		case it.IsActive && contains(filters, "active"):
+			fallthrough
+		case !it.IsActive && contains(filters, "system"):
+			fallthrough
+		case (it.IsRoot || it.IsAdmin) && contains(filters, "admin"):
+			filtered = append(filtered, it)
+		}
+	}
+
+	return filtered
+}
+
+// AddToGroup adds a User to a preexisting group. Returns whether or not the modification ran, along with an error in
+// case something went wrong
+func (c *Context) AddToGroup(user *User, group string) (ran bool, err error) {
+	if !contains(user.Groups, group) {
+		cmd := exec.Command("usermod", "-aG", group, user.Name)
+		if err = cmd.Run(); err != nil {
+			return false, err
+		}
+
+		user.Groups = append(user.Groups, group)
+		ran = true
+	}
+	return ran, err
+}
```

Comment:
Maybe the following instead?

``` go
func (c *Context) AddToGroup(user *User, group string) (ran bool, err error) {
	if contains(user.Groups, group) {
		return
	}
	
	cmd := exec.Command("usermod", "-aG", group, user.Name)
	ran = true
	if err = cmd.Run(); err == nil {
		user.Groups = append(user.Groups, group)
	}

	return
}
```

Reply:
This would change the behavior of the function due to the placement of `ran = true`, but this doesn't actually affect anything due to how the control flow is set up, so this is fine.